{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1>snRNA-seq re-analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# original raw data acquired by Schirmer et al. (Nature, 2019)\n",
    "# NCBI Sequence Read Archive (SRA), accession number PRJNA544731\n",
    "\n",
    "# load libraries\n",
    "library(tidyverse)\n",
    "library(Seurat)\n",
    "library(Polychrome)\n",
    "\n",
    "# import UMI matrix from Cell Ranger\n",
    "matrix_dir = \"XXX\" # insert path to UMI matrix here\n",
    "data_raw = Read10X(data.dir = matrix_dir)\n",
    "data = CreateSeuratObject(counts = data_raw, project = \"snRNA-seq\", min.cells = 3, min.features = 200)\n",
    "\n",
    "# calculate the percentage of mitochondrial features\n",
    "mito.features = grep(pattern = \"^MT-\", x = rownames(x = data), value = TRUE)\n",
    "percent.mito = Matrix::colSums(x = GetAssayData(object = data, slot = 'counts')[mito.features, ]) / Matrix::colSums(x = GetAssayData(object = data, slot = 'counts'))\n",
    "data[['percent.mito']] = percent.mito # store information\n",
    "\n",
    "# filter out cells that have unique feature counts over 5000 or less than 200\n",
    "data = subset(data, subset = nFeature_RNA > 500 & nCount_RNA > 1000)\n",
    "\n",
    "# Normalizing the data\n",
    "data = NormalizeData(data, normalization.method = \"LogNormalize\", scale.factor = 1e4)\n",
    "\n",
    "# detection of variable genes across the single cells\n",
    "data <- FindVariableFeatures(data, selection.method = 'mean.var.plot', mean.cutoff = c(0.0125, 3), dispersion.cutoff = c(0.5, Inf))\n",
    "\n",
    "# remove MT genes\n",
    "markers.remove <- grep(pattern = \"^MT-\",  x = rownames(data), value = TRUE)\n",
    "\n",
    "data@assays$RNA@var.features <- data@assays$RNA@var.features[!(data@assays$RNA@var.features%in%markers.remove)]\n",
    "\n",
    "# scaling the data and removing unwanted sources of variation\n",
    "data = ScaleData(data)\n",
    "\n",
    "# perform linear dimensional reduction\n",
    "data = RunPCA(data, features = VariableFeatures(data), verbose = FALSE)\n",
    "# examine and visualize PCA results a few different ways\n",
    "print(x = data[['pca']], dims = 1:5, nfeatures = 5, projected = FALSE)\n",
    "\n",
    "# get batches based on cell names\n",
    "samples_batches = sapply(colnames(GetAssayData(data, slot = \"counts\")),\n",
    "                      FUN=function(x){substr(x,18,19)})\n",
    "\n",
    "# turn to numbers and add cell names to them\n",
    "samples_batches = as.numeric(as.factor(samples_batches))\n",
    "names(samples_batches) = colnames(GetAssayData(data, slot = \"counts\"))\n",
    "\n",
    "# plot genes per cell\n",
    "complexity.per.cell = apply(GetAssayData(data, slot = \"counts\"),\n",
    "                             2, function(x) sum(x>0))\n",
    "plot(complexity.per.cell ~ jitter(samples_batches,2))\n",
    "\n",
    "sample.effect = samples_batches\n",
    "data = AddMetaData(data, sample.effect, \"sample.effect\")\n",
    "\n",
    "for (i in 1:length(x = data.list)) {\n",
    "    data.list[[i]] = NormalizeData(object = data.list[[i]], verbose = FALSE)\n",
    "    data.list[[i]] = FindVariableFeatures(object = data.list[[i]], \n",
    "        selection.method = \"vst\", nfeatures = 2000, verbose = FALSE)\n",
    "}\n",
    "\n",
    "reference.list <- data.list\n",
    "data.anchors <- FindIntegrationAnchors(object.list = reference.list, dims = 1:30)\n",
    "\n",
    "data.integrated = IntegrateData(anchorset = data.anchors, dims = 1:30)\n",
    "\n",
    "# switch to integrated assay\n",
    "DefaultAssay(object = data.integrated) <- \"integrated\"\n",
    "\n",
    "# Scaling the data and removing unwanted sources of variation\n",
    "data.integrated <- ScaleData(object = data.integrated)\n",
    "\n",
    "# perform linear dimensional reduction\n",
    "data.integrated <- RunPCA(object = data.integrated, npcs = 30, verbose = FALSE)\n",
    "# examine and visualize PCA results a few different ways\n",
    "print(x = data.integrated[['pca']], dims = 1:5, nfeatures = 5, projected = FALSE)\n",
    "VizDimLoadings(object = data.integrated, dims = 1:2)\n",
    "DimPlot(object = data.integrated)\n",
    "DimPlot(object = data.integrated, group.by=\"sample.effect\")\n",
    "\n",
    "# dimensional reduction and clustering\n",
    "data = FindNeighbors(data, dims = 1:16)\n",
    "data = FindClusters(data, resolution = 0.8)\n",
    "\n",
    "# perform non-linear dimensional reduction (tSNE)\n",
    "data = RunTSNE(data, dims = 1:16)\n",
    "\n",
    "# visualize data\n",
    "data(palette36) # color palette for clusters\n",
    "colpal = palette36[1:33]\n",
    "colpal = colpal %>% as.vector()\n",
    "\n",
    "# tSNE plot and clustering\n",
    "Idents(data) <- \"RNA_snn_res.0.8\"\n",
    "DimPlot(data, reduction = 'tsne', cols = colpal, label = F, label.size = 5, pt.size = 1) + \n",
    "  theme_void() + NoLegend()\n",
    "ggsave('output/tsne.png', plot = last_plot(), scale = 1, dpi = 900)\n",
    "DimPlot(data, reduction = 'tsne', cols = colpal, label = T, label.size = 5, pt.size = 1) + \n",
    "  theme_void()\n",
    "ggsave('output/tsne.pdf', plot = last_plot(), scale = 1)\n",
    "\n",
    "# signature genes for clusters (as used by Schirmer et al. for cell type identification)\n",
    "sig_genes = c('SYT1', 'SLC17A7', 'GAD2', 'AQP4', 'PLP1', 'P2RY12', 'CSF1R', 'CSF1', 'IL34')\n",
    "for (i in 1:length(sig_genes)){\n",
    "    FeaturePlot(data, features = sig_genes[i], label.size = 5, cols = c(\"lightgrey\", \"red3\"), pt.size = 1.5) +\n",
    "        theme_void() + NoLegend() + theme(title = element_blank())\n",
    "    ggsave(paste0('output/', 'tsne_', tolower(sig_genes[i]), '.png'), plot = last_plot(), scale = 1, dpi = 900, bg = \"transparent\")\n",
    "}\n",
    "\n",
    "# CSF1R, CSF1 and IL34 in tsne\n",
    "goi = c('CSF1R', 'IL34', 'CSF1') # define genes of interest\n",
    "for (i in 1:length(goi)){\n",
    "    FeaturePlot(data, features = goi[i], label.size = 5, cols = c(\"lightgrey\", \"red3\"), pt.size = 1.5) +\n",
    "        theme_void() + NoLegend() + theme(title = element_blank())\n",
    "    ggsave(paste0('output/', 'tsne_', tolower(goi[i]), '.png'), plot = last_plot(), scale = 1, dpi = 900)\n",
    "}\n",
    "\n",
    "# cluster-specific expression of CSF1R, CSF1 and IL34\n",
    "goi = c('CSF1R', 'IL34', 'CSF1') # define genes of interest\n",
    "for (i in 1:length(goi)){\n",
    "    Idents(data) <- \"RNA_snn_res.0.8\"\n",
    "    DotPlot(data, features = goi[i], dot.scale = 3, cols = c('red3', 'red3'), split.by = \"diagnosis\") +\n",
    "      theme(text=element_text(size=6,  family=\"sans\"),\n",
    "        axis.text.x = element_text(angle=45, vjust=1, hjust=1, size=6, family=\"sans\"),\n",
    "        axis.text.y = element_text(size=6, family=\"sans\", hjust = 1, vjust = 0.35),\n",
    "        axis.line = element_line(colour = 'black', size = 0.2),\n",
    "        axis.ticks = element_line(colour = 'black', size = 0.2)\n",
    "        )\n",
    "    ggsave(paste0('output/', 'dotplot_', tolower(goi[i]), '_msctrl.pdf'), plot = last_plot(), width = 6, height = 16, units = \"cm\")\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.0.2",
   "language": "R",
   "name": "ir32"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
